### ISSUE TITLE
[NLL] Bad higher ranked subtype error

### ISSUE THREAD

**Issue by matthewjasper** [rust-lang/rust#57374]
With the removal of the leak check the MIR type checker is now responsible for reporting higher-ranked lifetime errors in full NLL mode. The error messages are not currently very helpful, since they weren't user visible until now.

The following code

```rust
#![feature(nll)]

fn main() {
    let x: fn(&'static ()) = |_| {};
    let y: for<'a> fn(&'a ()) = x;
}
```

gives the following output

```text
error: higher-ranked subtype error
  --> <source>:19:12
   |
19 |     let x: fn(&'static ()) = |_| {};
   |            ^^^^^^^^^^^^^^^
error: aborting due to previous error
Compiler returned: 1
```

In migrate mode or with AST borrowck the error is much clearer:

```text
error[E0308]: mismatched types
  --> <source>:20:33
   |
20 |     let y: for<'a> fn(&'a ()) = x;
   |                                 ^ one type is more general than the other
   |
   = note: expected type `for<'a> fn(&'a ())`
              found type `fn(&'static ())`
error: aborting due to previous error
For more information about this error, try `rustc --explain E0308`.
Compiler returned: 1
```

cc @rust-lang/wg-compiler-nll

**Comment by nikomatsakis** [comment]
Yeah, I intended to port the newer errors to NLL but didn't do it yet since they are hidden by migration mode. But clearly we need to do that before we can stabilize NLL further.

**Comment by nikomatsakis** [comment]
Nominating for discussion in the NLL meeting -- this should probably be prioritized.

**Comment by nikomatsakis** [comment]
Discussed in the NLL meeting. Assigning to @lqd but also to me to leave some mentoring notes and/or sync with @lqd

**Comment by pnkfelix** [comment]
N.B.: This bug continues to persist even after Issue #57362 was resolved by PR #57901

**Comment by lqd** [comment]
With the temporary return of the leak-check from #58592, this issue has expectedly "un-regressed" to diagnostics levels very similar to pre-Universes or AST borrowck:

```
error[E0308]: mismatched types
 --> src/main.rs:5:33
  |
5 |     let y: for<'a> fn(&'a ()) = x;
  |                                 ^ expected concrete lifetime, found bound lifetime parameter 'a
  |
  = note: expected type `for<'a> fn(&'a ())`
             found type `fn(&'static ())`
```

**Comment by pnkfelix** [comment]
As discussed in the NLL meeting last night, one can use `-Z no-leak-check` to skip the leak check and thus return to the (regressed) behavior and thus attempt to fix it properly.

**Comment by lqd** [comment]
As mentioned on Zulip, I have made some progress here (and have notes from the exploratory analysis), but will need some guidance to continue, so unassigning myself until then.

**Comment by pnkfelix** [comment]
triage: downgrading to P-medium.

Blocks: #59490

**Comment by nnethercote** [comment]
I hit this today working on the compiler, calling a function with this signature:
```
pub fn link_name<F>(check_name: F, attrs: &[ast::Attribute]) -> Option<Symbol>
where
    F: Fn(&ast::Attribute, Symbol) -> bool
```
I had no idea what "Bad higher ranked subtype error" meant, so I googled and found this issue, and concluded it was somehow related to lifetimes. After some experimentation I got this signature to work:
```
pub fn link_name<'a, F>(check_name: F, attrs: &'a [ast::Attribute]) -> Option<Symbol>
where
    F: Fn(&'a ast::Attribute, Symbol) -> bool
```
So it was a happy ending, but there was some wailing and gnashing of teeth along the way.

**Comment by nikomatsakis** [comment]
Oof, I'm surprised you hit it actually

**Comment by nnethercote** [comment]
I hit it again a couple of days later in a different piece of code, but I didn't record the exact problem that time.

If the error message said something like "there might be a problem with lifetimes of a closure", that would be a significant hint in the right direction.

**Comment by camelid** [comment]
I've been hitting this in rustdoc recently (cc #79781).

**Comment by Aaron1011** [comment]
@lqd: Now that https://github.com/rust-lang/rust/pull/86700, what still needs to be done here?

**Comment by lqd** [comment]
There are still a few cases where "higher-ranked subtype error"s are emitted in their original terse style, including cases of type ascription like in the OP (but these exact samples don't trigger it anymore). Since the non-NLL errors are still better than the NLL ones here, we will still need to reach parity there in order to not degrade diagnostics. It's not as bad as before but not as good as migrate mode yet.

I'll update the tracking issue with all the remaining work I'm aware of once I'm back from vacation (and there's surely more work to-be-done that I don't know about yet).

**Comment by lqd** [comment]
None that I'm personally aware of.

Note that the TAIT test triggering the `higher-ranked subtype error` is really triggering it, whereas the polonius ones were likely fixed by the 2 previous PRs: these blessed expectations haven't been updated yet (and CI doesn't validate them as of yet).

This issue is probably the closest/only issue tracking this specific diagnostic, so we could still use it to track the rest of the work that they need, or close it and open new ones.

From #86700 you can see that there are at least a few remaining tasks:
- matching non-NLL output in these error messages, as requested by @estebank [here](https://github.com/rust-lang/rust/pull/86700#discussion_r664789685). This probably will require passing these obligations and causes around. Otherwise we won't be able to be as precise as the current output and it would regress the diagnostics too much. There are a number of these cases Esteban pointed out during review.
- the parity to non-NLL is also needed in some of the names currently used in the HRTB errors, for example [this](https://github.com/rust-lang/rust/blob/84b018341284798fa47f8171f4eb04f18f45cf23/src/test/ui/hr-subtype/hr-subtype.bound_a_b_ret_a_vs_bound_a_ret_a.nll.stderr) uses synthetic names for the regions, whereas [non-NLL output](https://github.com/rust-lang/rust/blob/84b018341284798fa47f8171f4eb04f18f45cf23/src/test/ui/hr-subtype/hr-subtype.bound_a_b_ret_a_vs_bound_a_ret_a.stderr) uses the names from user code. It's likely that fixing the point above and having the correct obligations would also fix this though.
(- some [clean-up of the implementation](https://github.com/rust-lang/rust/pull/86700#discussion_r689086688) is also required, but that's more of a task for enabling full-NLLs rather than this issue's errors)

**Comment by Aaron1011** [comment]
> matching non-NLL output in these error messages, as requested by @estebank here. This probably will require passing these obligations and causes around. Otherwise we won't be able to be as precise as the current output and it would regress the diagnostics too much. There are a number of these cases Esteban pointed out during review.

I'll start working on that

**Comment by Aaron1011** [comment]
I've opened https://github.com/rust-lang/rust/pull/89249 to improve the cause information for some NLL error messages.

**Comment by Aaron1011** [comment]
> the parity to non-NLL is also needed in some of the names currently used in the HRTB errors, for example this uses synthetic names for the regions, whereas non-NLL output uses the names from user code. It's likely that fixing the point above and having the correct obligations would also fix this though.

I've opened https://github.com/rust-lang/rust/pull/89250 to address this. It turned out to be caused by bound region anonimization during typeck.

**Comment by Aaron1011** [comment]
Progress update:

> From #86700 you can see that there are at least a few remaining tasks:

> matching non-NLL output in these error messages, as requested by @estebank here. This probably will require passing these obligations and causes around. Otherwise we won't be able to be as precise as the current output and it would regress the diagnostics too much. There are a number of these cases Esteban pointed out during review.

PR https://github.com/rust-lang/rust/pull/89249 has been merged, which makes NLL mode display the `note` that we display in non-NLL mode. I left comments on @estebank 's other requests in that PR - both of them are issues in non-NLL mode as well, so they're not directly related to this issue.

> the parity to non-NLL is also needed in some of the names currently used in the HRTB errors, for example this uses synthetic names for the regions, whereas non-NLL output uses the names from user code. It's likely that fixing the point above and having the correct obligations would also fix this though.

This was addressed in PR https://github.com/rust-lang/rust/pull/89250

> (some clean-up of the implementation is also required, but that's more of a task for enabling full-NLLs rather than this issue's errors)

This still needs to be done, but it's not currently affecting the user-visible output

**Comment by Aaron1011** [comment]
Are there any other user-visible diagnostic issues that need addressing?

**Comment by lqd** [comment]
On stable: I don't _think_ so. It's also unlikely there are other open issues about them.

On nightly: like last time, some TAIT uses trigger the bad higher ranked subtype errors. Since Oli is currently revamping that, and it's unstable, there's no urgency to handle these IMO.

**Comment by Aaron1011** [comment]
Issue https://github.com/rust-lang/rust/issues/89620 hits a higher-ranked subtype error with `#![feature(nll)]`:

```rust
#![feature(nll)]

fn bar<Input>() -> impl Fn(Input) {
    |i| todo!()
}
    
fn foo() -> impl Fn(&str) {
    bar()
}
```

produces:

```
error: higher-ranked subtype error
 --> src/lib.rs:8:5
  |
8 |     bar()
  |     ^^^^^

error: could not compile `playground` due to previous error
```

**Comment by Aaron1011** [comment]
I've opened https://github.com/rust-lang/rust/pull/92306 to address the issue with opaque types on stable

**Comment by marmeladema** [comment]
The error message is now identical between stable and nightly with nll enabled:
```
error[E0308]: mismatched types
 --> src/main.rs:5:33
  |
5 |     let y: for<'a> fn(&'a ()) = x;
  |                                 ^ one type is more general than the other
  |
  = note: expected fn pointer `for<'a> fn(&'a ())`
             found fn pointer `fn(&())`

For more information about this error, try `rustc --explain E0308`.

Maybe this issue can now be closed?
```

**Comment by estebank** [comment]
The current output for [aaron's case above](https://github.com/rust-lang/rust/issues/57374#issuecomment-997282040)

With nll
```
error: implementation of `Fn` is not general enough
 --> src/lib.rs:8:5
  |
8 |     bar()
  |     ^^^^^ implementation of `Fn` is not general enough
  |
  = note: `impl Fn(&'2 str)` must implement `Fn<(&'1 str,)>`, for any lifetime `'1`...
  = note: ...but it actually implements `Fn<(&'2 str,)>`, for some specific lifetime `'2`

error[E0308]: mismatched types
 --> src/lib.rs:8:5
  |
3 | fn bar<Input>() -> impl Fn(Input) {
  |                    --------------
  |                    |
  |                    the expected opaque type
  |                    the found opaque type
...
8 |     bar()
  |     ^^^^^ one type is more general than the other
  |
  = note: expected associated type `<impl Fn(&str) as FnOnce<(&str,)>>::Output`
             found associated type `<impl Fn(&str) as FnOnce<(&str,)>>::Output`
```
and without:
```
error: implementation of `Fn` is not general enough
 --> src/lib.rs:7:27
  |
7 |   fn foo() -> impl Fn(&str) {
  |  ___________________________^
8 | |     bar()
9 | | }
  | |_^ implementation of `Fn` is not general enough
  |
  = note: `impl Fn(&'2 str)` must implement `Fn<(&'1 str,)>`, for any lifetime `'1`...
  = note: ...but it actually implements `Fn<(&'2 str,)>`, for some specific lifetime `'2`
```

**Comment by Aaron1011** [comment]
I think the `error: implementation of `Fn` is not general enough` message is better with NLL, since we point to the location responsible for producing the hidden type, instead of pointing at the entire function. However, the `mismatched types` error is confusing and redundant.

**Comment by jackh726** [comment]
Found a very similar `higher-ranked lifetime error` in https://github.com/rust-lang/rust/blob/master/src/test/ui/higher-rank-trait-bounds/issue-59311.nll.stderr

**Comment by estebank** [comment]
I think that

```
  = note: `impl Fn(&'2 str)` must implement `Fn<(&'1 str,)>`, for any lifetime `'1`...
  = note: ...but it actually implements `Fn<(&'2 str,)>`, for some specific lifetime `'2`
```

could be improved by making the output closer to code people can write

```
  = note: `impl Fn(&'2 str)` has a specific lifetime `'2`, but must work for `for<'1> Fn<(&'1 str,)>`
```

**Comment by jackh726** [comment]
@estebank I think you're correct, but I think that might be better for a separate issue?

**Comment by estebank** [comment]
@jackh726 yes, a separate PR is always in the table for these kind of fixes :)

**Comment by jackh726** [comment]
Imo this isn't the right issue for that. I think this should be purely to remove any "bad higher ranked subtype"-like errors. As it stands, nearly all have been removed and *match* non-NLL behavior. A separate issue could be opened to *improve* that error in NLL (unlikely to be relevant for non-NLL, since it'll likely get stabilized soon).
